### ISSUE TITLE
[Regression] Accidentally broke compilation with App Extensions / 1.9.7

### ISSUE THREAD

**Issue by martin-gearzero** [facebook/AsyncDisplayKit#1309]
Just wanted to make sure this is on the radar as I couldn't find another related ticket.

Currently master/1.9.7 fails compilation for App Extensions because it relies on PINCache. Specifically, [PINDiskCache start] uses UIApplication which is unavailable to app extensions.

I found [this](https://github.com/pinterest/PINCache/issues/37) closed ticket in PINCache [issues](https://github.com/pinterest/PINCache/issues?q=is%3Aissue+is%3Aclosed) which suggests that PINCache is not planning to support extensions.

Is the plan to drop support for extensions with the next release of ASDK? Personally, I find having that ability very convenient because I can share code between the two targets.

Thanks!

**Comment by appleguy** [comment]
@martin-gearzero thanks for reporting this.  We are actually not seeing this issue at Pinterest - we include AsyncDisplayKit in the app extension.  We'll definitely fix this as soon as we can find the cause.  Any other details (full Xcode logs / etc) would be ideal.  We may end up needing to build a test project to at least confirm whatever fix works.

**Comment by martin-gearzero** [comment]
@appleguy I played around with this and it may be a CocoaPods configuration issue on my end but I'm not positive. I've created a simple project to illustrate the problem [here](https://github.com/martin-gearzero/asdk-1309).

Specifically, if I have a simple Podfile to include AsyncDisplayKit 1.9.7 then there is no issue. However, I have separate target definitions in my [Podfile](https://github.com/martin-gearzero/asdk-1309/blob/master/Podfile) because I only need a subset of pods in my sharing extension.

I'm afraid I'm not too familiar with the inner workings of CocoaPods so I'm not sure how to proceed at this point. Is this something which needs to be fixed with ASDK/PINCache or is it a limitation of CocoaPods? Or perhaps it simply a misconfiguration in my project?

**Comment by chrisdanford** [comment]
PINCache (and PINRemoteImage) expect `PIN_APP_EXTENSIONS` to be defined when they're built for an extension.  The uses of UIApplication are guarded by:
```
#if !defined(PIN_APP_EXTENSIONS) && __IPHONE_OS_VERSION_MIN_REQUIRED >= __IPHONE_4_0
```
In our Podfile, we use the post_install hook to set define that flag.
```
post_install do |installer|
    target = installer.pods_project.targets.find { |t| t.to_s == "Pods-ShareExtension-PINCache" }

    if (target)
        target.build_configurations.each do |config|
            config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] = ['$(inherited)', 'PIN_APP_EXTENSIONS=1']
        end
    else
        puts "WARNING: Pods-ShareExtension-PINCache target not found"
    end

    target = installer.pods_project.targets.find { |t| t.to_s == "Pods-ShareExtension-PINRemoteImage" }

    if (target)
        target.build_configurations.each do |config|
            config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] = ['$(inherited)', 'PIN_APP_EXTENSIONS=1']
        end
    else
        puts "WARNING: Pods-ShareExtension-PINRemoteImage target not found"
    end
```

This isn't a great solution, but it appears there's no standard preprocessor flag to detect whether or not you're being compiled as part of an extension (http://stackoverflow.com/a/29611274)

**Comment by garrettmoon** [comment]
There's discussion around a possible solution in PINCache happening here: https://github.com/pinterest/PINCache/pull/68

**Comment by martin-gearzero** [comment]
@garrettmoon thanks for the heads up. Happy to see it looks like the issue is resolved.

@chrisdanford thanks for the workaround in the meantime. That did the trick!

**Comment by garrettmoon** [comment]
@martin-gearzero this is now fixed in 2.2.2 of PINCache. You should no longer need the post_install step and macro. A simple pod update (if you haven't specified a specific version of PINCache) should resolve.

**Comment by martin-gearzero** [comment]
Brilliant! Thanks @garrettmoon 

Closing as this appears to be resolved.
